src/crystal/system/win32/socket.cr

```diff
@@ -0,0 +1,425 @@
+
+  private def system_bind(addr, addrstr)
+    unless LibC.bind(fd, addr, addr.size) == 0
+      yield ::Socket::BindError.from_errno("Could not bind to '#{addrstr}'")
+    end
+  end
+
+  private def system_listen(backlog)
+    unless LibC.listen(fd, backlog) == 0
+      yield ::Socket::Error.from_errno("Listen failed")
+    end
+  end
+
+  protected def system_accept : Handle?
+    client_socket = create_handle(family, type, protocol, blocking)
+    initialize_handle(client_socket)
+
+    if system_accept(client_socket)
+      client_socket
+    else
+      LibC.closesocket(client_socket)
+
+      nil
+    end
+  end
+
+  protected def system_accept(client_socket : Handle) : Bool
+    address_size = sizeof(LibC::SOCKADDR_STORAGE) + 16
+    buffer_size = 0
+    output_buffer = Bytes.new(address_size * 2 + buffer_size)
+
+    success = overlapped_accept(fd, "AcceptEx") do |overlapped|
+      received_bytes = uninitialized UInt32
+
+      result = Crystal::System::Socket.accept_ex.call(fd, client_socket,
+        output_buffer.to_unsafe.as(Void*), buffer_size.to_u32!,
+        address_size.to_u32!, address_size.to_u32!, pointerof(received_bytes), overlapped)
+
+      if result.zero?
+        error = WinError.wsa_value
+
+        unless error.wsa_io_pending?
+          return false
+        end
+      end
+    end
+
+    return false unless success
+
+    # AcceptEx does not automatically set the socket options on the accepted
+    # socket to match those of the listening socket, we need to ask for that
+    # explicitly with SO_UPDATE_ACCEPT_CONTEXT
+    system_setsockopt client_socket, LibC::SO_UPDATE_ACCEPT_CONTEXT, fd
+
+    true
+  end
+
+  private def wsa_buffer(bytes)
+    wsabuf = LibC::WSABUF.new
+    wsabuf.len = bytes.size
+    wsabuf.buf = bytes.to_unsafe
+    wsabuf
+  end
+
+  private def system_send(message : Bytes) : Int32
+    wsabuf = wsa_buffer(message)
+
+    bytes = overlapped_write(fd, "WSASend") do |overlapped|
+      LibC.WSASend(fd, pointerof(wsabuf), 1, out bytes_sent, 0, overlapped, nil)
+    end
+
+    bytes.to_i32
+  end
+
+  private def system_send_to(bytes : Bytes, addr : ::Socket::Address)
+    wsabuf = wsa_buffer(bytes)
+    bytes_sent = overlapped_write(fd, "WSASendTo") do |overlapped|
+      LibC.WSASendTo(fd, pointerof(wsabuf), 1, out bytes_sent, 0, addr, addr.size, overlapped, nil)
+    end
+    raise ::Socket::Error.from_errno("Error sending datagram to #{addr}") if bytes_sent == -1
+
+    # to_i32 is fine because string/slice sizes are an Int32
+    bytes_sent.to_i32
+  end
+
+  private def system_receive(bytes)
+    sockaddr = Pointer(LibC::SOCKADDR_STORAGE).malloc.as(LibC::Sockaddr*)
+    # initialize sockaddr with the initialized family of the socket
+    copy = sockaddr.value
+    copy.sa_family = family
+    sockaddr.value = copy
+
+    addrlen = sizeof(LibC::SOCKADDR_STORAGE)
+
+    wsabuf = wsa_buffer(bytes)
+
+    flags = 0_u32
+    bytes_read = overlapped_read(fd, "WSARecvFrom") do |overlapped|
```

Comment:
As I understand it, the reason for the loop is mostly because with libevent the method is called twice in an async operation. The first time errors with `EAGAIN`. That error code is the only way for the loop in `evented_*` to repeat. There is no such thing in WSA.

Reply:
That may explain `evented_read` but it definitely don't explain `evented_write`

Reply:
The Winsock2 documentation does not mention anything about not consuming the entire buffer and having to re-issue an operation again. I'm pretty sure completion is only signaled when the read/write is performed entirely.

Reply:
Right. I guess people will complain if something doesn't work :D

Reply:
I'm sure this won't cover everything 😆